{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 1\n",
    "Write a function whose input is a string. This function determines the data type of the input string. The data types can be a float, int, or string. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_data_type(value):\n",
    "    \"\"\"\n",
    "    The function takes a string input and determines its data type to be either a float, int, or string. \n",
    "    \"\"\"    \n",
    "    try:\n",
    "        int(value)\n",
    "        return int\n",
    "    except:\n",
    "        pass\n",
    "    \n",
    "    try:\n",
    "        float(value)\n",
    "        return float\n",
    "    except:\n",
    "        return str"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 2\n",
    "Write a function whose input is a list of strings.  This function determines the correct data type of all the elements in the list. For example, ['1', '2', '3'] is int, ['1.1', '2.2', '3.3'] is float, ['1.1', '2', '3.3'] is also float, and ['1.1', '234String', '3.3'] is str. The purpose of this function to figure out what to cast an array of strings to. Some lists might be all ints, in which case the data type is int. Some might be a mixure of ints and floats, in which case the data type will be float. Some lists might be a mixture of ints, floats, and strings, in which case the data type of the list will be string. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_data_type_of_list(values):\n",
    "    \"\"\"\n",
    "    Function takes a list of strings and determines their data type. \n",
    "\n",
    "    \"\"\"\n",
    "    f_counter = 0\n",
    "    \n",
    "    for x in values:\n",
    "        if determine_data_type(x) == str:\n",
    "            return str\n",
    "        elif determine_data_type(x) == float:\n",
    "            f_counter += 1\n",
    "        else:\n",
    "            pass\n",
    "    \n",
    "    if f_counter >= 1:\n",
    "        return float\n",
    "    else:\n",
    "        return int"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 3\n",
    "Write a function whose inputs a format field and a sample field. The format field looks like `format_field = 'GT:AD:DP:GQ:PGT:PID:PL'` and\n",
    "the sample field looks like\n",
    "\n",
    "```\n",
    "sample_field = {'XG102': '1/1:0,76:76:99:1|1:48306945_C_G:3353,229,0',\n",
    "             'XG103': '1/1:0,52:52:99:.:.:1517,156,0',\n",
    "             'XG104': '0/1:34,38:72:99:.:.:938,0,796',\n",
    "             'XG202': '1/1:0,76:76:99:1|1:48306945_C_G:3353,229,0',\n",
    "             'XG203': '1/1:0,52:52:99:.:.:1517,156,0',\n",
    "             'XG204': '0/1:34,38:72:99:.:.:938,0,796',\n",
    "             'XG302': '1/1:0,76:76:99:1|1:48306945_C_G:3353,229,0',\n",
    "             'XG303': '1/1:0,52:52:99:.:.:1517,156,0',\n",
    "             'XG304': '0/1:34,38:72:99:.:.:938,0,796',\n",
    "             'XG402': '1/1:0,76:76:99:1|1:48306945_C_G:3353,229,0',\n",
    "             'XG403': '1/1:0,52:52:99:.:.:1517,156,0',\n",
    "             'XG404': '0/1:34,38:72:99:.:.:938,0,796'}\n",
    "```\n",
    "Transform the inputs such that the output looks like this:\n",
    "```\n",
    "output = {'XG102': {'AD': '0,76',\n",
    "           'DP': '76',\n",
    "           'GQ': '99',\n",
    "           'GT': '1/1',\n",
    "           'PGT': '1|1',\n",
    "           'PID': '48306945_C_G',\n",
    "           'PL': '3353,229,0'},\n",
    " 'XG103': {'AD': '0,52',\n",
    "           'DP': '52',\n",
    "           'GQ': '99',\n",
    "           'GT': '1/1',\n",
    "           'PGT': '.',\n",
    "           'PID': '.',\n",
    "           'PL': '1517,156,0'},\n",
    " 'XG104': {'AD': '34,38',\n",
    "           'DP': '72',\n",
    "           'GQ': '99',\n",
    "           'GT': '0/1',\n",
    "           'PGT': '.',\n",
    "           'PID': '.',\n",
    "           'PL': '938,0,796'},\n",
    " 'XG202': {'AD': '0,76',\n",
    "           'DP': '76',\n",
    "           'GQ': '99',\n",
    "           'GT': '1/1',\n",
    "           'PGT': '1|1',\n",
    "           'PID': '48306945_C_G',\n",
    "           'PL': '3353,229,0'},\n",
    " 'XG203': {'AD': '0,52',\n",
    "           'DP': '52',\n",
    "           'GQ': '99',\n",
    "           'GT': '1/1',\n",
    "           'PGT': '.',\n",
    "           'PID': '.',\n",
    "           'PL': '1517,156,0'},\n",
    " 'XG204': {'AD': '34,38',\n",
    "           'DP': '72',\n",
    "           'GQ': '99',\n",
    "           'GT': '0/1',\n",
    "           'PGT': '.',\n",
    "           'PID': '.',\n",
    "           'PL': '938,0,796'},\n",
    " 'XG302': {'AD': '0,76',\n",
    "           'DP': '76',\n",
    "           'GQ': '99',\n",
    "           'GT': '1/1',\n",
    "           'PGT': '1|1',\n",
    "           'PID': '48306945_C_G',\n",
    "           'PL': '3353,229,0'},\n",
    " 'XG303': {'AD': '0,52',\n",
    "           'DP': '52',\n",
    "           'GQ': '99',\n",
    "           'GT': '1/1',\n",
    "           'PGT': '.',\n",
    "           'PID': '.',\n",
    "           'PL': '1517,156,0'},\n",
    " 'XG304': {'AD': '34,38',\n",
    "           'DP': '72',\n",
    "           'GQ': '99',\n",
    "           'GT': '0/1',\n",
    "           'PGT': '.',\n",
    "           'PID': '.',\n",
    "           'PL': '938,0,796'},\n",
    " 'XG402': {'AD': '0,76',\n",
    "           'DP': '76',\n",
    "           'GQ': '99',\n",
    "           'GT': '1/1',\n",
    "           'PGT': '1|1',\n",
    "           'PID': '48306945_C_G',\n",
    "           'PL': '3353,229,0'},\n",
    " 'XG403': {'AD': '0,52',\n",
    "           'DP': '52',\n",
    "           'GQ': '99',\n",
    "           'GT': '1/1',\n",
    "           'PGT': '.',\n",
    "           'PID': '.',\n",
    "           'PL': '1517,156,0'},\n",
    " 'XG404': {'AD': '34,38',\n",
    "           'DP': '72',\n",
    "           'GQ': '99',\n",
    "           'GT': '0/1',\n",
    "           'PGT': '.',\n",
    "           'PID': '.',\n",
    "           'PL': '938,0,796'}}\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "format_field = \"GT:AD:DP:GQ:PGT:PID:PL\"\n",
    "sample_field = {'XG102': '1/1:0,76:76:99:1|1:48306945_C_G:3353,229,0',\n",
    "             'XG103': '1/1:0,52:52:99:.:.:1517,156,0',\n",
    "             'XG104': '0/1:34,38:72:99:.:.:938,0,796',\n",
    "             'XG202': '1/1:0,76:76:99:1|1:48306945_C_G:3353,229,0',\n",
    "             'XG203': '1/1:0,52:52:99:.:.:1517,156,0',\n",
    "             'XG204': '0/1:34,38:72:99:.:.:938,0,796',\n",
    "             'XG302': '1/1:0,76:76:99:1|1:48306945_C_G:3353,229,0',\n",
    "             'XG303': '1/1:0,52:52:99:.:.:1517,156,0',\n",
    "             'XG304': '0/1:34,38:72:99:.:.:938,0,796',\n",
    "             'XG402': '1/1:0,76:76:99:1|1:48306945_C_G:3353,229,0',\n",
    "             'XG403': '1/1:0,52:52:99:.:.:1517,156,0',\n",
    "             'XG404': '0/1:34,38:72:99:.:.:938,0,796'}\n",
    "\n",
    "\n",
    "def format_sample_fields(format_field, sample_field):\n",
    "    \"\"\"\n",
    "    Format the sample fields given the description above. Data is already provided.\n",
    "    \"\"\"    \n",
    "    solution={}\n",
    "    \n",
    "    for keys in sample_field:\n",
    "        solution[keys] = dict(zip(format_field.split(':'), sample_field[keys].split(':')))\n",
    "    \n",
    "    return solution    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 4\n",
    "Write a function whose inputs are a list containing the vcf header and a variant line.  The function should return a dictionary using the header as keys and the variant line as values. The function should use the `format_sample_fields` you wrote previously to format the sample fields. The output of the first line looks like this:\n",
    "```\n",
    "{'ALT': 'G',\n",
    " 'CHROM': '4',\n",
    " 'FILTER': 'PASS',\n",
    " 'ID': '.',\n",
    " 'INFO': 'AC=1;AF=0.167;AN=6;BaseQRankSum=-2.542;ClippingRankSum=0;DP=180;ExcessHet=3.0103;FS=0;MLEAC=1;MLEAF=0.167;MQ=52.77;MQRankSum=-4.631;QD=0.39;ReadPosRankSum=1.45;SOR=0.758;VQSLOD=-8.209;culprit=MQ;ANNOVAR_DATE=2018-04-16;Func.refGene=intergenic;Gene.refGene=IL2\\\\x3bIL21;GeneDetail.refGene=dist\\\\x3d38536\\\\x3bdist\\\\x3d117597;ExonicFunc.refGene=.;AAChange.refGene=.;Func.ensGene=intergenic;Gene.ensGene=ENSG00000109471\\\\x3bENSG00000138684;GeneDetail.ensGene=dist\\\\x3d38306\\\\x3bdist\\\\x3d117597;ExonicFunc.ensGene=.;AAChange.ensGene=.;cytoBand=4q27;gwasCatalog=.;tfbsConsSites=.;wgRna=.;targetScanS=.;Gene_symbol=.;OXPHOS_Complex=.;Ensembl_Gene_ID=.;Ensembl_Protein_ID=.;Uniprot_Name=.;Uniprot_ID=.;NCBI_Gene_ID=.;NCBI_Protein_ID=.;Gene_pos=.;AA_pos=.;AA_sub=.;Codon_sub=.;dbSNP_ID=.;PhyloP_46V=.;PhastCons_46V=.;PhyloP_100V=.;PhastCons_100V=.;SiteVar=.;PolyPhen2_prediction=.;PolyPhen2_score=.;SIFT_prediction=.;SIFT_score=.;FatHmm_prediction=.;FatHmm_score=.;PROVEAN_prediction=.;PROVEAN_score=.;MutAss_prediction=.;MutAss_score=.;EFIN_Swiss_Prot_Score=.;EFIN_Swiss_Prot_Prediction=.;EFIN_HumDiv_Score=.;EFIN_HumDiv_Prediction=.;CADD_score=.;CADD_Phred_score=.;CADD_prediction=.;Carol_prediction=.;Carol_score=.;Condel_score=.;Condel_pred=.;COVEC_WMV=.;COVEC_WMV_prediction=.;PolyPhen2_score_transf=.;PolyPhen2_pred_transf=.;SIFT_score_transf=.;SIFT_pred_transf=.;MutAss_score_transf=.;MutAss_pred_transf=.;Perc_coevo_Sites=.;Mean_MI_score=.;COSMIC_ID=.;Tumor_site=.;Examined_samples=.;Mutation_frequency=.;US=.;Status=.;Associated_disease=.;Presence_in_TD=.;Class_predicted=.;Prob_N=.;Prob_P=.;SIFT_score=.;SIFT_converted_rankscore=.;SIFT_pred=.;Polyphen2_HDIV_score=.;Polyphen2_HDIV_rankscore=.;Polyphen2_HDIV_pred=.;Polyphen2_HVAR_score=.;Polyphen2_HVAR_rankscore=.;Polyphen2_HVAR_pred=.;LRT_score=.;LRT_converted_rankscore=.;LRT_pred=.;MutationTaster_score=.;MutationTaster_converted_rankscore=.;MutationTaster_pred=.;MutationAssessor_score=.;MutationAssessor_score_rankscore=.;MutationAssessor_pred=.;FATHMM_score=.;FATHMM_converted_rankscore=.;FATHMM_pred=.;PROVEAN_score=.;PROVEAN_converted_rankscore=.;PROVEAN_pred=.;VEST3_score=.;VEST3_rankscore=.;MetaSVM_score=.;MetaSVM_rankscore=.;MetaSVM_pred=.;MetaLR_score=.;MetaLR_rankscore=.;MetaLR_pred=.;M-CAP_score=.;M-CAP_rankscore=.;M-CAP_pred=.;CADD_raw=.;CADD_raw_rankscore=.;CADD_phred=.;DANN_score=.;DANN_rankscore=.;fathmm-MKL_coding_score=.;fathmm-MKL_coding_rankscore=.;fathmm-MKL_coding_pred=.;Eigen_coding_or_noncoding=.;Eigen-raw=.;Eigen-PC-raw=.;GenoCanyon_score=.;GenoCanyon_score_rankscore=.;integrated_fitCons_score=.;integrated_fitCons_score_rankscore=.;integrated_confidence_value=.;GERP++_RS=.;GERP++_RS_rankscore=.;phyloP100way_vertebrate=.;phyloP100way_vertebrate_rankscore=.;phyloP20way_mammalian=.;phyloP20way_mammalian_rankscore=.;phastCons100way_vertebrate=.;phastCons100way_vertebrate_rankscore=.;phastCons20way_mammalian=.;phastCons20way_mammalian_rankscore=.;SiPhy_29way_logOdds=.;SiPhy_29way_logOdds_rankscore=.;Interpro_domain=.;GTEx_V6_gene=.;GTEx_V6_tissue=.;esp6500siv2_all=.;esp6500siv2_aa=.;esp6500siv2_ea=.;ExAC_ALL=.;ExAC_AFR=.;ExAC_AMR=.;ExAC_EAS=.;ExAC_FIN=.;ExAC_NFE=.;ExAC_OTH=.;ExAC_SAS=.;ExAC_nontcga_ALL=.;ExAC_nontcga_AFR=.;ExAC_nontcga_AMR=.;ExAC_nontcga_EAS=.;ExAC_nontcga_FIN=.;ExAC_nontcga_NFE=.;ExAC_nontcga_OTH=.;ExAC_nontcga_SAS=.;ExAC_nonpsych_ALL=.;ExAC_nonpsych_AFR=.;ExAC_nonpsych_AMR=.;ExAC_nonpsych_EAS=.;ExAC_nonpsych_FIN=.;ExAC_nonpsych_NFE=.;ExAC_nonpsych_OTH=.;ExAC_nonpsych_SAS=.;1000g2015aug_all=.;1000g2015aug_afr=.;1000g2015aug_amr=.;1000g2015aug_eur=.;1000g2015aug_sas=.;CLNALLELEID=.;CLNDN=.;CLNDISDB=.;CLNREVSTAT=.;CLNSIG=.;dbscSNV_ADA_SCORE=.;dbscSNV_RF_SCORE=.;snp138NonFlagged=.;avsnp150=.;CADD13_RawScore=0.015973;CADD13_PHRED=2.741;Eigen=-0.3239;REVEL=.;MCAP=.;Interpro_domain=.;ICGC_Id=.;ICGC_Occurrence=.;gnomAD_genome_ALL=0.0003;gnomAD_genome_AFR=0.0001;gnomAD_genome_AMR=0;gnomAD_genome_ASJ=0;gnomAD_genome_EAS=0.0007;gnomAD_genome_FIN=0.0009;gnomAD_genome_NFE=0.0002;gnomAD_genome_OTH=0.0011;gerp++gt2=.;cosmic70=.;InterVar_automated=.;PVS1=.;PS1=.;PS2=.;PS3=.;PS4=.;PM1=.;PM2=.;PM3=.;PM4=.;PM5=.;PM6=.;PP1=.;PP2=.;PP3=.;PP4=.;PP5=.;BA1=.;BS1=.;BS2=.;BS3=.;BS4=.;BP1=.;BP2=.;BP3=.;BP4=.;BP5=.;BP6=.;BP7=.;Kaviar_AF=.;Kaviar_AC=.;Kaviar_AN=.;ALLELE_END',\n",
    " 'POS': '123416186',\n",
    " 'QUAL': '23.25',\n",
    " 'REF': 'A',\n",
    " 'SAMPLE': {'XG102': {'AD': '51,8',\n",
    "            'DP': '59',\n",
    "            'GQ': '32',\n",
    "            'GT': '0/1',\n",
    "            'PL': '32,0,1388'},\n",
    "       'XG103': {'AD': '47,0',\n",
    "            'DP': '47',\n",
    "            'GQ': '99',\n",
    "            'GT': '0/0',\n",
    "            'PL': '0,114,1353'},\n",
    "       'XG104': {'AD': '74,0',\n",
    "            'DP': '74',\n",
    "            'GQ': '51',\n",
    "            'GT': '0/0',\n",
    "            'PL': '0,51,1827'},\n",
    "       'XG202': {'AD': '51,8',\n",
    "            'DP': '59',\n",
    "            'GQ': '32',\n",
    "            'GT': '0/1',\n",
    "            'PL': '32,0,1388'},\n",
    "       'XG203': {'AD': '47,0',\n",
    "            'DP': '47',\n",
    "            'GQ': '99',\n",
    "            'GT': '0/0',\n",
    "            'PL': '0,114,1353'},\n",
    "       'XG204': {'AD': '74,0',\n",
    "            'DP': '74',\n",
    "            'GQ': '51',\n",
    "            'GT': '0/0',\n",
    "            'PL': '0,51,1827'},\n",
    "       'XG302': {'AD': '51,8',\n",
    "            'DP': '59',\n",
    "            'GQ': '32',\n",
    "            'GT': '0/1',\n",
    "            'PL': '32,0,1388'},\n",
    "       'XG303': {'AD': '47,0',\n",
    "            'DP': '47',\n",
    "            'GQ': '99',\n",
    "            'GT': '0/0',\n",
    "            'PL': '0,114,1353'},\n",
    "       'XG304': {'AD': '74,0',\n",
    "            'DP': '74',\n",
    "            'GQ': '51',\n",
    "            'GT': '0/0',\n",
    "            'PL': '0,51,1827'},\n",
    "       'XG402': {'AD': '51,8',\n",
    "            'DP': '59',\n",
    "            'GQ': '32',\n",
    "            'GT': '0/1',\n",
    "            'PL': '32,0,1388'},\n",
    "       'XG403': {'AD': '47,0',\n",
    "            'DP': '47',\n",
    "            'GQ': '99',\n",
    "            'GT': '0/0',\n",
    "            'PL': '0,114,1353'},\n",
    "       'XG404': {'AD': '74,0',\n",
    "            'DP': '74',\n",
    "            'GQ': '51',\n",
    "            'GT': '0/0',\n",
    "            'PL': '0,51,1827'}}}\n",
    "            \n",
    "```\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "header = \"CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tXG102\tXG103\tXG104\tXG202\tXG203\tXG204\tXG302\tXG303\tXG304\tXG402\tXG403\tXG404\".split('\\t')\n",
    "line = \"\"\"4\t123416186\t.\tA\tG\t23.25\tPASS\tAC=1;AF=0.167;AN=6;BaseQRankSum=-2.542;ClippingRankSum=0;DP=180;ExcessHet=3.0103;FS=0;MLEAC=1;MLEAF=0.167;MQ=52.77;MQRankSum=-4.631;QD=0.39;ReadPosRankSum=1.45;SOR=0.758;VQSLOD=-8.209;culprit=MQ;ANNOVAR_DATE=2018-04-16;Func.refGene=intergenic;Gene.refGene=IL2\\x3bIL21;GeneDetail.refGene=dist\\x3d38536\\x3bdist\\x3d117597;ExonicFunc.refGene=.;AAChange.refGene=.;Func.ensGene=intergenic;Gene.ensGene=ENSG00000109471\\x3bENSG00000138684;GeneDetail.ensGene=dist\\x3d38306\\x3bdist\\x3d117597;ExonicFunc.ensGene=.;AAChange.ensGene=.;cytoBand=4q27;gwasCatalog=.;tfbsConsSites=.;wgRna=.;targetScanS=.;Gene_symbol=.;OXPHOS_Complex=.;Ensembl_Gene_ID=.;Ensembl_Protein_ID=.;Uniprot_Name=.;Uniprot_ID=.;NCBI_Gene_ID=.;NCBI_Protein_ID=.;Gene_pos=.;AA_pos=.;AA_sub=.;Codon_sub=.;dbSNP_ID=.;PhyloP_46V=.;PhastCons_46V=.;PhyloP_100V=.;PhastCons_100V=.;SiteVar=.;PolyPhen2_prediction=.;PolyPhen2_score=.;SIFT_prediction=.;SIFT_score=.;FatHmm_prediction=.;FatHmm_score=.;PROVEAN_prediction=.;PROVEAN_score=.;MutAss_prediction=.;MutAss_score=.;EFIN_Swiss_Prot_Score=.;EFIN_Swiss_Prot_Prediction=.;EFIN_HumDiv_Score=.;EFIN_HumDiv_Prediction=.;CADD_score=.;CADD_Phred_score=.;CADD_prediction=.;Carol_prediction=.;Carol_score=.;Condel_score=.;Condel_pred=.;COVEC_WMV=.;COVEC_WMV_prediction=.;PolyPhen2_score_transf=.;PolyPhen2_pred_transf=.;SIFT_score_transf=.;SIFT_pred_transf=.;MutAss_score_transf=.;MutAss_pred_transf=.;Perc_coevo_Sites=.;Mean_MI_score=.;COSMIC_ID=.;Tumor_site=.;Examined_samples=.;Mutation_frequency=.;US=.;Status=.;Associated_disease=.;Presence_in_TD=.;Class_predicted=.;Prob_N=.;Prob_P=.;SIFT_score=.;SIFT_converted_rankscore=.;SIFT_pred=.;Polyphen2_HDIV_score=.;Polyphen2_HDIV_rankscore=.;Polyphen2_HDIV_pred=.;Polyphen2_HVAR_score=.;Polyphen2_HVAR_rankscore=.;Polyphen2_HVAR_pred=.;LRT_score=.;LRT_converted_rankscore=.;LRT_pred=.;MutationTaster_score=.;MutationTaster_converted_rankscore=.;MutationTaster_pred=.;MutationAssessor_score=.;MutationAssessor_score_rankscore=.;MutationAssessor_pred=.;FATHMM_score=.;FATHMM_converted_rankscore=.;FATHMM_pred=.;PROVEAN_score=.;PROVEAN_converted_rankscore=.;PROVEAN_pred=.;VEST3_score=.;VEST3_rankscore=.;MetaSVM_score=.;MetaSVM_rankscore=.;MetaSVM_pred=.;MetaLR_score=.;MetaLR_rankscore=.;MetaLR_pred=.;M-CAP_score=.;M-CAP_rankscore=.;M-CAP_pred=.;CADD_raw=.;CADD_raw_rankscore=.;CADD_phred=.;DANN_score=.;DANN_rankscore=.;fathmm-MKL_coding_score=.;fathmm-MKL_coding_rankscore=.;fathmm-MKL_coding_pred=.;Eigen_coding_or_noncoding=.;Eigen-raw=.;Eigen-PC-raw=.;GenoCanyon_score=.;GenoCanyon_score_rankscore=.;integrated_fitCons_score=.;integrated_fitCons_score_rankscore=.;integrated_confidence_value=.;GERP++_RS=.;GERP++_RS_rankscore=.;phyloP100way_vertebrate=.;phyloP100way_vertebrate_rankscore=.;phyloP20way_mammalian=.;phyloP20way_mammalian_rankscore=.;phastCons100way_vertebrate=.;phastCons100way_vertebrate_rankscore=.;phastCons20way_mammalian=.;phastCons20way_mammalian_rankscore=.;SiPhy_29way_logOdds=.;SiPhy_29way_logOdds_rankscore=.;Interpro_domain=.;GTEx_V6_gene=.;GTEx_V6_tissue=.;esp6500siv2_all=.;esp6500siv2_aa=.;esp6500siv2_ea=.;ExAC_ALL=.;ExAC_AFR=.;ExAC_AMR=.;ExAC_EAS=.;ExAC_FIN=.;ExAC_NFE=.;ExAC_OTH=.;ExAC_SAS=.;ExAC_nontcga_ALL=.;ExAC_nontcga_AFR=.;ExAC_nontcga_AMR=.;ExAC_nontcga_EAS=.;ExAC_nontcga_FIN=.;ExAC_nontcga_NFE=.;ExAC_nontcga_OTH=.;ExAC_nontcga_SAS=.;ExAC_nonpsych_ALL=.;ExAC_nonpsych_AFR=.;ExAC_nonpsych_AMR=.;ExAC_nonpsych_EAS=.;ExAC_nonpsych_FIN=.;ExAC_nonpsych_NFE=.;ExAC_nonpsych_OTH=.;ExAC_nonpsych_SAS=.;1000g2015aug_all=.;1000g2015aug_afr=.;1000g2015aug_amr=.;1000g2015aug_eur=.;1000g2015aug_sas=.;CLNALLELEID=.;CLNDN=.;CLNDISDB=.;CLNREVSTAT=.;CLNSIG=.;dbscSNV_ADA_SCORE=.;dbscSNV_RF_SCORE=.;snp138NonFlagged=.;avsnp150=.;CADD13_RawScore=0.015973;CADD13_PHRED=2.741;Eigen=-0.3239;REVEL=.;MCAP=.;Interpro_domain=.;ICGC_Id=.;ICGC_Occurrence=.;gnomAD_genome_ALL=0.0003;gnomAD_genome_AFR=0.0001;gnomAD_genome_AMR=0;gnomAD_genome_ASJ=0;gnomAD_genome_EAS=0.0007;gnomAD_genome_FIN=0.0009;gnomAD_genome_NFE=0.0002;gnomAD_genome_OTH=0.0011;gerp++gt2=.;cosmic70=.;InterVar_automated=.;PVS1=.;PS1=.;PS2=.;PS3=.;PS4=.;PM1=.;PM2=.;PM3=.;PM4=.;PM5=.;PM6=.;PP1=.;PP2=.;PP3=.;PP4=.;PP5=.;BA1=.;BS1=.;BS2=.;BS3=.;BS4=.;BP1=.;BP2=.;BP3=.;BP4=.;BP5=.;BP6=.;BP7=.;Kaviar_AF=.;Kaviar_AC=.;Kaviar_AN=.;ALLELE_END\tGT:AD:DP:GQ:PL\t0/1:51,8:59:32:32,0,1388\t0/0:47,0:47:99:0,114,1353\t0/0:74,0:74:51:0,51,1827\t0/1:51,8:59:32:32,0,1388\t0/0:47,0:47:99:0,114,1353\t0/0:74,0:74:51:0,51,1827\t0/1:51,8:59:32:32,0,1388\t0/0:47,0:47:99:0,114,1353\t0/0:74,0:74:51:0,51,1827\t0/1:51,8:59:32:32,0,1388\t0/0:47,0:47:99:0,114,1353\t0/0:74,0:74:51:0,51,1827\"\"\"\n",
    "\n",
    "def create_dict_from_line(header, line):\n",
    "    \"\"\"\n",
    "    Given the header and a single line, transform them into dictionary as described above. \n",
    "    Header and line input are provided in this cell. \n",
    "    \"\"\"    \n",
    "    sample_field_1 = dict(zip(header[9:],line.split('\\t')[9:]))\n",
    "    format_field_1 = line.split('\\t')[8]\n",
    "    \n",
    "    solution = dict(zip(header[:8],line.split('\\t')[:8]))\n",
    "    \n",
    "    solution['SAMPLE'] = dict(format_sample_fields(format_field_1, sample_field_1))\n",
    "    \n",
    "    return solution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 5\n",
    "Write a function whose input is a filename for a vcf. The function reads the vcf file one variant at a time and transforms it into a dictionary using the create_dict_from_line function. It returns a list containing all the variant dictionaries. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf_file(filename):\n",
    "    \"\"\"\n",
    "    See description above\n",
    "    \"\"\"    \n",
    "    variant_list = []\n",
    "    \n",
    "    with open(filename, 'r') as f:\n",
    "        file_data = f.readlines()\n",
    "    \n",
    "    filtered_data = [element for element in file_data if not element.startswith('##')]\n",
    "    \n",
    "    header = filtered_data[0].strip().strip('#').split('\\t')\n",
    "    \n",
    "    for i in range(1,len(filtered_data)):\n",
    "        line = filtered_data[i].strip()\n",
    "        variant_list.append(create_dict_from_line(header, line))\n",
    "        \n",
    "    return variant_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 6\n",
    "Write a function that extracts the info field from the data dictionary that was created in the previous part. The function should return all the info field dictionaries as list. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_info_field(data):\n",
    "    \"\"\"\n",
    "    See description in part 6\n",
    "    \"\"\"    \n",
    "    return [data[0]['INFO']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 7\n",
    "You now need to figure out that data types for each of the info fields. \n",
    "Begin by writing a function that first takes the info fields and turns them into a dictionary. Make sure to skip any fields that do not have a value or are missing a value. Also replace \\\\x3b with a comma and \\\\x3d with an equal sign. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_dictionary_of_info_field_values(data):\n",
    "    \"\"\"\n",
    "    See description of part 7\n",
    "    \"\"\"\n",
    "    keys = []\n",
    "    values = []\n",
    "    for x in data[0].split(';'):\n",
    "        try:\n",
    "            val = (x.split('=')[1].replace('\\\\x3b',',').replace('\\\\x3d','=')).strip('^\\.$')\n",
    "            values.append(val)\n",
    "            key = x.split('=')[0].replace('\\\\x3b',',').replace('\\\\x3d','=')\n",
    "            keys.append(key)\n",
    "        except:\n",
    "            pass\n",
    "    \n",
    "    output = dict(zip(keys,([x] if x else [] for x in values)))\n",
    "    \n",
    "    return output\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 8\n",
    "Write a function whose input is the output from `create_dictionary_of_info_field_values` and uses the previously written function `determine_data_type_of_list` to determine the data type of each of the info fields. The output is a dictionary whose keys are the name of the info fields and values are the data type. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_data_type_of_info_fields(data):\n",
    "    \"\"\"\n",
    "    See desription in part 8\n",
    "    \"\"\"\n",
    "    output = {}\n",
    "    \n",
    "    for key in data:\n",
    "        output[key] = determine_data_type_of_list(data[key])\n",
    "    \n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 9\n",
    "Write a function whose first input is the data from  `read_vcf_file` and the second input is the output from `determine_data_type_of_info_fields`. The function converts the info field into a dictionary and uses the data types that you determined to cast each field into the correct data type. Make sure to convert the `POS` to int and `QUAL` to float. Replace any `\\\\x3b` with a comma and any `\\\\x3d` with an equal sign. The output will look something like this (I have removed most of the fields):\n",
    "\n",
    "The output will look something like this \n",
    "```\n",
    "{\n",
    "        \"ALT\": \"G\",\n",
    "        \"CHROM\": \"4\",\n",
    "        \"FILTER\": \"PASS\",\n",
    "        \"ID\": \".\",\n",
    "        \"INFO\": {\n",
    "           \n",
    "            \"Gene.ensGene\": \"ENSG00000109471,ENSG00000138684\",\n",
    "            \"Gene.refGene\": \"IL2,IL21\",\n",
    "            \"GeneDetail.ensGene\": \"dist=38306,dist=117597\",\n",
    "            \"GeneDetail.refGene\": \"dist=38536,dist=117597\"\n",
    "        },\n",
    "        \"POS\": 123416186,\n",
    "        \"QUAL\" :23.25,\n",
    "        \"REF\": \"A\",\n",
    "        \"SAMPLE\": {\n",
    "            \"XG102\": {\n",
    "                \"AD\": \"51,8\",\n",
    "                \"DP\": \"59\",\n",
    "                \"GQ\": \"32\",\n",
    "                \"GT\": \"0/1\",\n",
    "                \"PL\": \"32,0,1388\"\n",
    "            }\n",
    "    }\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def format_data(data, info_field_data_type):\n",
    "    \"\"\"\n",
    "    See description 9\n",
    "    \"\"\"\n",
    "    in_dict = data[0]\n",
    "    empty_key_list = []\n",
    "    info_dict = create_dictionary_of_info_field_values([data[0]['INFO']])\n",
    "    \n",
    "    for keys in info_dict:\n",
    "        if info_dict[keys] != []:\n",
    "            info_dict[keys] = info_field_data_type[keys](','.join(info_dict[keys]))\n",
    "        else:\n",
    "            info_dict[keys] = ''\n",
    "            empty_key_list.append(keys)\n",
    "        \n",
    "    for x in empty_key_list:\n",
    "        del info_dict[x]\n",
    "    \n",
    "    in_dict['INFO'] = info_dict\n",
    "    \n",
    "    in_dict['POS'] = int(in_dict['POS'])\n",
    "    \n",
    "    in_dict['QUAL'] = float(in_dict['QUAL'])\n",
    "    \n",
    "    return in_dict"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 10\n",
    "Write a function whose inputs are a Python dictionary and filename. The function will saves the dictionary as a json file using the filename given. Use the json library. Use these options to correctly format your JSON -- `sort_keys=True, indent=4, separators=(',', ': ')`. \n",
    "Use this function to save your parsed data as a json file. \n",
    "\n",
    "USE THE FOLLOWING FILENAME TO SAVE YOUR FILE: `lab1.json`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "def save_data_as_json(data,filename):\n",
    "    with open(filename,'w') as out_file:\n",
    "        json.dump(data, out_file, sort_keys=True, indent=4, separators=(',', ': '))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 11\n",
    "Write a function whose input is a filename for a json file. The function should use the filename to read the JSON file in which you saved your final parsed data. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "def load_data_from_json(filename):\n",
    "    with open(filename) as json_file:\n",
    "        return json.load(json_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 12\n",
    "Write a function whose inputs are `CHROM`, `REF`, `ALT`, `POS`, and `filename`.  Using these inputs, the function should load a JSON file using the given filename and return a list of variants that match \n",
    "the given CHROM, REF, ALT, and POS. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_variant(CHROM, REF, ALT, POS, filename):\n",
    "    \n",
    "    json_list = []\n",
    "    result_list = []\n",
    "    \n",
    "    data = read_vcf_file('lab1_data.vcf')\n",
    "    \n",
    "    for variants in data:\n",
    "        info_field_data = extract_info_field([variants])\n",
    "        info_field_list = create_dictionary_of_info_field_values(info_field_data)\n",
    "        info_field_data_type = determine_data_type_of_info_fields(info_field_list)\n",
    "        data = format_data([variants], info_field_data_type)\n",
    "        json_list.append(data)\n",
    "    \n",
    "    save_data_as_json(json_list, filename)\n",
    "    data_loaded = load_data_from_json(filename)\n",
    "    \n",
    "    for variants_dict in data_loaded:\n",
    "        if variants_dict['CHROM'] == CHROM and variants_dict['REF'] == REF and variants_dict['ALT'] == ALT and variants_dict['POS'] == POS:\n",
    "            result_list.append(variants_dict)\n",
    "    \n",
    "    return result_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'ALT': 'G',\n",
       "  'CHROM': '13',\n",
       "  'FILTER': 'PASS',\n",
       "  'ID': 'rs4421887',\n",
       "  'INFO': {'1000g2015aug_afr': 0.8313,\n",
       "   '1000g2015aug_all': 0.952676,\n",
       "   '1000g2015aug_amr': 0.9841,\n",
       "   '1000g2015aug_eur': 1,\n",
       "   '1000g2015aug_sas': 0.9969,\n",
       "   'AC': 6,\n",
       "   'AF': 1,\n",
       "   'AN': 6,\n",
       "   'ANNOVAR_DATE': '2018-04-16',\n",
       "   'BaseQRankSum': 1.77,\n",
       "   'CADD13_PHRED': 3.712,\n",
       "   'CADD13_RawScore': 0.109684,\n",
       "   'ClippingRankSum': 0,\n",
       "   'DP': 176,\n",
       "   'Eigen': -0.7945,\n",
       "   'ExcessHet': 3.0103,\n",
       "   'FS': 0,\n",
       "   'Func.ensGene': 'intergenic',\n",
       "   'Func.refGene': 'intergenic',\n",
       "   'Gene.ensGene': 'ENSG00000264387,ENSG00000228611',\n",
       "   'Gene.refGene': 'MIR5007,PRR20D',\n",
       "   'GeneDetail.ensGene': 'dist=543620,dist=281031',\n",
       "   'GeneDetail.refGene': 'dist=543620,dist=1422749',\n",
       "   'Kaviar_AC': 3,\n",
       "   'Kaviar_AF': 0.0001153,\n",
       "   'Kaviar_AN': 26028,\n",
       "   'MLEAC': 6,\n",
       "   'MLEAF': 1,\n",
       "   'MQ': 60,\n",
       "   'MQRankSum': 0,\n",
       "   'QD': 28.95,\n",
       "   'ReadPosRankSum': -0.338,\n",
       "   'SOR': 0.257,\n",
       "   'VQSLOD': 22.53,\n",
       "   'avsnp150': 'rs4421887',\n",
       "   'culprit': 'MQ',\n",
       "   'cytoBand': '13q21.1',\n",
       "   'gnomAD_genome_AFR': 0.8475,\n",
       "   'gnomAD_genome_ALL': 0.9546,\n",
       "   'gnomAD_genome_AMR': 0.9843,\n",
       "   'gnomAD_genome_ASJ': 0.9834,\n",
       "   'gnomAD_genome_EAS': 1,\n",
       "   'gnomAD_genome_FIN': 0.9951,\n",
       "   'gnomAD_genome_NFE': 0.9987,\n",
       "   'gnomAD_genome_OTH': 0.9887,\n",
       "   'snp138NonFlagged': 'rs4421887'},\n",
       "  'POS': 56292303,\n",
       "  'QUAL': 5037.69,\n",
       "  'REF': 'T',\n",
       "  'SAMPLE': {'XG102': {'AD': '0,62',\n",
       "    'DP': '62',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '1796,185,0'},\n",
       "   'XG103': {'AD': '0,41',\n",
       "    'DP': '41',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '1218,122,0'},\n",
       "   'XG104': {'AD': '1,70',\n",
       "    'DP': '71',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '2037,202,0'},\n",
       "   'XG202': {'AD': '0,62',\n",
       "    'DP': '62',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '1796,185,0'},\n",
       "   'XG203': {'AD': '0,41',\n",
       "    'DP': '41',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '1218,122,0'},\n",
       "   'XG204': {'AD': '1,70',\n",
       "    'DP': '71',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '2037,202,0'},\n",
       "   'XG302': {'AD': '0,62',\n",
       "    'DP': '62',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '1796,185,0'},\n",
       "   'XG303': {'AD': '0,41',\n",
       "    'DP': '41',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '1218,122,0'},\n",
       "   'XG304': {'AD': '1,70',\n",
       "    'DP': '71',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '2037,202,0'},\n",
       "   'XG402': {'AD': '0,62',\n",
       "    'DP': '62',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '1796,185,0'},\n",
       "   'XG403': {'AD': '0,41',\n",
       "    'DP': '41',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '1218,122,0'},\n",
       "   'XG404': {'AD': '1,70',\n",
       "    'DP': '71',\n",
       "    'GQ': '99',\n",
       "    'GT': '1/1',\n",
       "    'PL': '2037,202,0'}}}]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "find_variant(\"13\", \"T\", \"G\", 56292303, 'lab1.json')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
